### PR TITLE
Implement deep merge

### PR DESCRIPTION
**Pull request by dpc**
This is an alternative approach suggested in #120 

TODO:

* [x] seems like due to using traits, I don't need all that extra type tracking - remove
* [x] validate in `rustshop` code
* [x] cleanup spaces (why do you use tabs for Rust projects, BTW? :P)
* [x] revise some TODOs left in the code

### CODE REVIEW

File: k8s-openapi-codegen-common/templates/int_or_string.rs [hunk 1]
```diff
@@ -65,4 +67,4 @@ impl {local}serde::Serialize for {type_name} {{
             {type_name}::String(s) => s.serialize(serializer),
         }}
     }}
-}}
```
Comment (Arnavion) [inline k8s-openapi-codegen-common/templates/int_or_string.rs:68]: Revert these whitespace changes. You can see the whitespace errors this has introduced in the generated code.

File: src/deep_merge.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                Entry::Occupied(mut e) => { e.get_mut().merge_from(v); },
+            }
+        }
+    }
+}
+
+impl<T> DeepMerge for Option<T> where T: DeepMerge {
+    fn merge_from(&mut self, other: Self) {
+        if let Some(other) = other {
+            if let Some(ref mut s) = *self {
```
Comment (Arnavion) [inline src/deep_merge.rs:55]:
```suggestion
            if let Some(s) = self {
```

File: src/deep_merge.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        self.append(&mut other);
+    }
+}
+
+impl<K, V> DeepMerge for BTreeMap<K, V> where K : Ord, V : DeepMerge {
+    fn merge_from(&mut self, other: Self) {
+        for (k, v) in other.into_iter() {
+            match self.entry(k) {
+                Entry::Vacant(e) => { e.insert(v); },
+                Entry::Occupied(mut e) => { e.get_mut().merge_from(v); },
```
Comment (Arnavion) [inline src/deep_merge.rs:46]:
```suggestion
                Entry::Occupied(e) => { e.into_mut().merge_from(v); },
```

File: src/deep_merge.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+impl<'a> DeepMerge for &'a [std::string::String] {}
+
+impl<'a, T> DeepMerge for &'a T {}
+
+impl DeepMerge for String {}
+impl DeepMerge for crate::ByteString {}
+impl<T> DeepMerge for chrono::DateTime<T> where T : chrono::TimeZone {}
+
+/// TODO: well... actually we probably have to descend recursively into it or something :/
+impl DeepMerge for serde_json::Value {}
```
Comment (Arnavion) [inline src/deep_merge.rs:26]: https://stackoverflow.com/a/54118457/545475

File: k8s-openapi-codegen-common/templates/struct.rs [hunk 5]
```diff
@@ -1,2 +1,9 @@
 struct {type_name}{type_generics_type}{type_generics_where} {{
-{fields}}}
\ No newline at end of file
+{fields}
+}}
+
+impl{type_generics_type} crate::DeepMerge for {type_name}{type_generics_type} {type_generics_where} {{
```
Comment (Arnavion) [inline k8s-openapi-codegen-common/templates/struct.rs:5]:
```suggestion
impl{type_generics_type} {local}deep_merge::DeepMerge for {type_name}{type_generics_type} {type_generics_where} {{
```

File: src/lib.rs [hunk 6]
```diff
@@ -404,6 +404,9 @@ pub use serde_value;
 #[cfg(feature = "api")]
 pub use url;
 
+
+pub mod deep_merge;
```
Comment (Arnavion) [inline src/lib.rs:408]:
```suggestion
mod deep_merge;
```

The only useful thing in this mod is the trait.

File: k8s-openapi-codegen-common/templates/int_or_string.rs (hunk 1 above)
Comment (dpc) [reply k8s-openapi-codegen-common/templates/int_or_string.rs:68]: What kind of text editor are you using? [kakoune can't do it](https://github.com/mawww/kakoune/issues/2147), `nano` can't do it. `gedit` can't do it, `vim` can't do it ...

I'll start to look for hex editor ...

File: src/deep_merge.rs (hunk 4 above)
Comment (dpc) [reply src/deep_merge.rs:26]: Thx!

File: k8s-openapi-codegen-common/templates/int_or_string.rs (hunk 1 above)
Comment (dpc) [reply k8s-openapi-codegen-common/templates/int_or_string.rs:68]: Hex editor helped. :D

**Comment by dpc** [comment]
[I've rewrote the previous code I had to use this PR instead](https://github.com/rustshop/rustshop/commit/f41976c7a65986b78f07f22a028e8ff40991341a).

I had to import a macro for map literals, and with that it's not too bad. Lack of any auto-`Option` conversion/insertion is a noticeable and annoying difference, but that's kind of it. Though some parts are even better (setting multiple fields on the same struct).

Once (if) [`std::default`](https://doc.rust-lang.org/std/default/fn.default.html) stabilizes some of this `Default::default` noise can be removed, though for now I can always just use [`default::default`](https://docs.rs/default/latest/default/fn.default.html).

[`Option::get_or_insert_default`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_default) will be a minor improvement too once it stabilizes.

**Comment by Arnavion** [comment]
I've fixed the whitespace errors, missing impl for `serde_json::Value`, wrong reference to the trait in the templates, incorrect impls of the trait for borrowed types, etc etc in [pr/121.](https://github.com/Arnavion/k8s-openapi/compare/pr/121)

---

One thing that's broken is the custom derive, since the codegen of `FooBar: DeepMerge` requires `FooBarSpec: DeepMerge`, but that's a user type. I don't want to make users impl it by hand, at least not if they don't care for `FooBar: DeepMerge` in the first place.

So the options seem to be:

1. Make it easy for all users (both those who want `FooBar: DeepMerge` and those who don't care for it) to impl `DeepMerge` for `FooBarSpec` via a new custom derive. This is not straightforward because the `Spec` type could be anything, eg a struct or enum with other arbitrary types.

1. Emit `impl DeepMerge for FooBar where FooBarSpec: DeepMerge { ... }`, so that users who don't impl the trait for their `FooBarSpec` still have compilable code. But this requires `#![feature(trivial_bounds)]`, so it's a no-go.

1. Make the `impl DeepMerge for FooBar` codegen conditional on an attr of `#[derive(CustomResourceDefinition)]` so that only users who opt in to it get it.

(3) seems to be the only path forward.

**Comment by dpc** [comment]
Just a status update: I've had much less time recently, and it looks like it will not change for a month or more, so probably won't be able to improve it anytime soon. The issues you describe seem a bit out of depth for me RN, so I would have to invest quite a bit of time just to start working on them.

Feel free to take over if you can/wish. I probably will get back to it if needs more work, once I find some time.

**Comment by Arnavion** [comment]
Yes, it's on my plate. I've just been busy myself.

**Comment by Arnavion** [comment]
Okay, I've updated [pr/121](https://github.com/Arnavion/k8s-openapi/compare/pr/121) with the change to the custom derive to make the impl of `DeepMerge` for the generated CR type opt-in. If you're okay with that branch, I'll merge it.

The other major changes are:

- I removed the default impl of `merge_from` that did `*self = other;`, since I want to make sure users implementing the trait on their own types do it "properly" (field-by-field).

- I changed the codegen for structs to do `DeepMerge::merge_from(&mut self.field, other.field)` instead of `self.field.merge_from(other.field)` because the field type could have an inherent `fn merge_from` that would be chosen over the trait method. None of the structs in this crate have such an inherent method, but the user's CRD spec type could have one which would be problematic for the generated CR's impl.

- Changed the impl for `Vec` to do `self.extend(other);` instead of `self.append(&mut other);`. They both forward to the same specialized function internally, so `.extend` is clearer since it consumed `other` instead of draining it.

**Comment by dpc** [comment]
My code seem to compile as is with this change and I have very minimal requirements. Everything that you described and I understood LGTM. :)

**Pull request closed without merging**